### PR TITLE
Cover shopping list line items in integration test scenarios

### PR DESCRIPTION
**Pull request by ahmetoz**
All possible cases for the shopping lists are tested with the new integration tests. (except unexpected cases, for those unit tests are enough).

**Reviewer tips:**
- Don't check the diff, because I created all integration tests from the stratch, as before it was not testing the update actions fully. Check first externalsource package, then projectsource package.
 
- As line-items depend on the variant, the service was not expanding it. Check: (fetchMatchingShoppingListsByKeys) service method.

- custom update actions like AddLineItemWithSku were not deserializing the custom object, so updated.

-  order of anonymous id and customer action is changed, because of the scenario below:

````
{
        "version" : 1,
        "actions" : [ {
          "action" : "setCustomer",
          "customer" : {
            "typeId" : "customer",
            "id" : "04d1d86a-6387-45dd-b8a1-4075caf842c9"
          }
        }, {
          "action" : "setAnonymousId"
        } ]
      }

{
  "statusCode" : 400,
  "message" : "The resource was already claimed by a customer: 04d1d86a-6387-45dd-b8a1-4075caf842c9",
  "errors" : [ {
    "code" : "InvalidOperation",
    "message" : "The resource was already claimed by a customer: 04d1d86a-6387-45dd-b8a1-4075caf842c9",
    "action" : {
      "action" : "setAnonymousId"
    },
    "actionIndex" : 2
  } ]
}
````

### CODE REVIEW

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/shoppinglists/ShoppingListSyncIT.java [hunk 1]
```diff
@@ -110,202 +119,173 @@ void sync_WithSameShoppingList_ShouldNotUpdateCustomer() {
...
-        assertThat(exceptions).isEmpty();
-        assertThat(updateActionList).isEmpty();
-
-        assertThat(shoppingListSyncStatistics).hasValues(1, 1, 0, 0);
-
+                + "(0 created, 0 updated and 0 failed to sync).");
     }
 
     @Test
     void sync_WithModifiedShoppingList_ShouldUpdateShoppingList() {
```
Comment (lojzatran) [inline src/integration-test/java/com/commercetools/sync/integration/externalsource/shoppinglists/ShoppingListSyncIT.java:126]: Please add a comment or link to the example in ADR that this test is trying to address.

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/shoppinglists/ShoppingListSyncIT.java [hunk 2]
```diff
@@ -110,202 +119,173 @@ void sync_WithSameShoppingList_ShouldNotUpdateCustomer() {
...
-                .toCompletableFuture()
-                .join();
-
-        assertThat(errorMessages).isEmpty();
-        assertThat(warningMessages).isEmpty();
-        assertThat(exceptions).isEmpty();
-        assertThat(updateActionList).containsExactly(
-                SetCustomer.of(customer)
-        );
+        assertUpdateActions();
```
Comment (lojzatran) [inline src/integration-test/java/com/commercetools/sync/integration/externalsource/shoppinglists/ShoppingListSyncIT.java:138]: I would actually assert that the shoppingList was updated (as your test is named). You can leave the update actions assert here, but I think the focus of the integration test shouldn't be too much on the implementation, but rather if the result is achieved.

File: src/integration-test/java/com/commercetools/sync/integration/ctpprojectsource/shoppinglists/ShoppingListSyncIT.java [hunk 3]
```diff
@@ -89,167 +103,62 @@ void sync_WithoutUpdates_ShouldReturnProperStatistics() {
...
-        final List<ShoppingList> shoppingLists = CTP_SOURCE_CLIENT
-                .execute(buildShoppingListQuery())
-                .toCompletableFuture()
-                .join()
-                .getResults();
+        createSampleCustomerJaneDoe(CTP_SOURCE_CLIENT);
+        final Customer sampleCustomerJaneDoe = createSampleCustomerJaneDoe(CTP_TARGET_CLIENT);
 
-        final List<ShoppingListDraft> shoppingListDrafts = mapToShoppingListDrafts(shoppingLists);
+        final List<ShoppingListDraft> updatedShoppingListDrafts =
```
Comment (lojzatran) [inline src/integration-test/java/com/commercetools/sync/integration/ctpprojectsource/shoppinglists/ShoppingListSyncIT.java:128]: Do you also want to check the line items reordering here or just with one test?

**Review by lojzatran — CHANGES_REQUESTED** [review]

File: src/integration-test/java/com/commercetools/sync/integration/ctpprojectsource/shoppinglists/ShoppingListSyncIT.java (hunk 3 above)
Comment (ahmetoz) [reply src/integration-test/java/com/commercetools/sync/integration/ctpprojectsource/shoppinglists/ShoppingListSyncIT.java:128]: Would be good, but we have enough coverage, so imo not necessary.

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/shoppinglists/ShoppingListSyncIT.java (hunk 2 above)
Comment (ahmetoz) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/shoppinglists/ShoppingListSyncIT.java:138]: You're right, I am trusting here the update actions and statistics, also a bit black box testing. But could be added, let me check the effort and decide later.

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/shoppinglists/ShoppingListSyncIT.java (hunk 1 above)
Comment (ahmetoz) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/shoppinglists/ShoppingListSyncIT.java:129]: https://github.com/commercetools/commercetools-sync-java/pull/626/commits/d722e5dbb9cb5c557ce034b79bcafc71c7c58428

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/shoppinglists/ShoppingListSyncIT.java (hunk 2 above)
Comment (ahmetoz) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/shoppinglists/ShoppingListSyncIT.java:141]: https://github.com/commercetools/commercetools-sync-java/pull/626/commits/0729ca1f3b8e70c278e640b900d0b55d930edb2c

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/shoppinglists/ShoppingListSyncIT.java (hunk 1 above)
Comment (lojzatran) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/shoppinglists/ShoppingListSyncIT.java:129]: I was thinking more like a compare where you have old and new shopping lists, so I can see the diff:

```
oldShoppingList{
name: "abc"
[LineItem1, LineItem2]
}

newShoppingList{
name: "newAbc"
[LineItem1]
}
```

**Review by lojzatran — APPROVED** [review]
LGTM 🎉

File: src/integration-test/java/com/commercetools/sync/integration/ctpprojectsource/shoppinglists/ShoppingListSyncIT.java [hunk 4]
```diff
@@ -89,167 +103,62 @@ void sync_WithoutUpdates_ShouldReturnProperStatistics() {
...
         assertThat(shoppingListSyncStatistics
             .getReportMessage())
-            .isEqualTo("Summary: 1 shopping lists were processed in total "
-                    +  "(1 created, 0 updated and 0 failed to sync).");
+            .isEqualTo("Summary: 2 shopping lists were processed in total "
+                + "(1 created, 0 updated and 0 failed to sync).");
     }
 
     @Test
-    void sync_WithUpdatedShoppingList_ShouldReturnProperStatistics() {
```
[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/integration-test/java/com/commercetools/sync/integration/ctpprojectsource/shoppinglists/ShoppingListSyncIT.java:102]: `sync_WithUpdatedCustomerOnShoppingList_ShouldReturnProperStatistics` is covering an update for customer field, for other fields check the externalsource tests.

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/shoppinglists/ShoppingListSyncIT.java (hunk 1 above)
Comment (ahmetoz) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/shoppinglists/ShoppingListSyncIT.java:129]: would be better, but in the end, I am thinking the scenarios are clear.

File: src/integration-test/java/com/commercetools/sync/integration/ctpprojectsource/shoppinglists/ShoppingListSyncIT.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (leungkinghin) [reply src/integration-test/java/com/commercetools/sync/integration/ctpprojectsource/shoppinglists/ShoppingListSyncIT.java:102]: If so, can we add updated anonymous ID and slug for ShoppingList in line 134?

[the remark this answers is not in the document]
Comment (ahmetoz) [reply src/integration-test/java/com/commercetools/sync/integration/ctpprojectsource/shoppinglists/ShoppingListSyncIT.java:102]: Why not: https://github.com/commercetools/commercetools-sync-java/pull/626/commits/00b3c7fb59b6160f17402a2c6a1efb990dd01709

**Review by leungkinghin — APPROVED** [review]

**Pull request merged**